pkg/finalize_bundle.go

```diff
@@ -0,0 +1,188 @@
+
+	for _, img := range images {
+		name, err := image.NewName(img)
+		if err != nil {
+			fmt.Printf("err %v\n", err)
+		}
+		n := strings.ReplaceAll(name.String(), "/", "_")
+		bunImg := bundle.Image{}
+		d, err := r.Digest(name)
+		if err != nil {
+			return err
+		}
+		bunImg.Digest = d.String()
+		nameWithDigest, err := name.WithDigest(d)
+		if err != nil {
+			return err
+		}
+		bunImg.Image = nameWithDigest.String()
+		mfst.Images[n] = bunImg
+
+		replacements = append(replacements, img, nameWithDigest.String())
+	}
+
+	err = marshallJsonFile(bundlePath, mfst)
+	if err != nil {
+		return err
+	}
+
+	err = ReplaceInKabManifest(kabManifestPath, *strings.NewReplacer(replacements...))
+
+	return err
+
+}
+
+func unmarshallFile(path string, str interface{}) error {
+	mfstBytes, err := ioutil.ReadFile(path)
+	if err != nil {
+		fmt.Printf("error reading file %s: %v", path, err)
+		return err
+	}
+	err = yaml.Unmarshal(mfstBytes, str)
+	if err != nil {
+		fmt.Printf("error unmarshalling file %s: %v", path, err)
+		return err
+	}
+	return nil
+}
+
+func marshallJsonFile(path string, str interface{}) error {
+	mfstBytes, err := json.MarshalIndent(str, "", "    ")
+	if err != nil {
+		return err
+	}
+	err = writeFile(path, str, mfstBytes)
+	return err
+}
+
+func marshallYamlFile(path string, str interface{}) error {
+	mfstBytes, err := yaml.Marshal(str)
+	if err != nil {
+		return err
+	}
+	err = writeFile(path, str, mfstBytes)
+	return err
+}
+
+func writeFile(path string, str interface{}, content []byte) error {
+	err := ioutil.WriteFile(path, content, 0644)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("wrote file %s\n", path)
+	return nil
+}
+
+func GetImagesFromKabManifest(kabManifestPath string) ([]string, error) {
+	kabMfst := &v1alpha1.Manifest{}
+	err := unmarshallFile(kabManifestPath, kabMfst)
+	if err != nil {
+		return nil, err
+	}
+
+	images := []string{}
+
+	err = kabMfst.VisitResources(func(res v1alpha1.KabResource) error {
+		fmt.Fprintf(os.Stderr, "Scanning %s\n", res.Name)
+
+		var err error
+		var imgs []string
+		if res.Content != "" {
+			imgs, err = scan.ListSortedImagesFromContent([]byte(res.Content))
+		} else {
+			imgs, err = scan.ListSortedImagesFromKubernetesManifest(res.Path, "")
+		}
+		if err != nil {
+			return err
+		}
+
+		images = append(images, imgs...)
+
+		return nil
+	})
+	return images, nil
+}
+
+func InlineContentInKabManifest(kabManifestPath string) error {
+	kabMfst := &v1alpha1.Manifest{}
+	err := unmarshallFile(kabManifestPath, kabMfst)
+	if err != nil {
+		return err
+	}
+
+	err = kabMfst.PatchResourceContent(func(res *v1alpha1.KabResource) (string, error) {
+		contentBytes, err := furl.Read(res.Path, "")
+		if err != nil {
+			return "", err
+		}
+		return string(contentBytes), nil
+	})
+	if err != nil {
+		return err
+	}
+
+	err = marshallYamlFile(kabManifestPath, kabMfst)
+	return err
+}
+
+func ReplaceInKabManifest(kabManifestPath string, replacer strings.Replacer) error {
```

Comment:
why does every method load the manifest from disk and then write it back out? Could it not operate on the Manifest object directly?